lib/services/spatial.js

```diff
@@ -22,36 +24,127 @@ class SpatialService {
   }
 
 
-  _destroyLayers(fourfours, core, cb) {
-    return async.map(fourfours, core.destroy.bind(core), (err, destroyResponses) => {
+  _createOrReplace(core, layer) {
+    return (layer, cb) => {
+      if (layer.uid === Layer.EMPTY) {
+        return core.create(layer, (err, resp) => {
+          cb(err, [resp, false]);
+        });
+      }
+      return core.replace(layer, (err, resp) => {
+        cb(err, [resp, true]);
+      });
+    };
+  }
+
+  /**
+   * Helper for turning a list of layers into a flat list of
+   * [layerUid, column] tuples
+   */
+  _flatColSpecs(layers) {
+    return _.flatten(layers.map((layer) => {
+      return layer.columns.map((col) => [layer.uid, col]);
+    }), true);
+  }
+
+  _destroyLayers(layers, core, cb) {
+    return async.map(layers, core.destroy.bind(core), (err, destroyResponses) => {
```

Comment:
Change to `async.mapLimit` instead of `async.map` as you have in other map calls